python/private/pip_config_settings.bzl

```diff
@@ -0,0 +1,354 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module is used to construct the config settings for selecting which distribution is used in the pip hub repository.
+
+Bazel's selects work by selecting the most-specialized configuration setting
+that matches the target platform. We can leverage this fact to ensure that the
+most specialized wheels are used by default with the users being able to
+configure string_flag values to select the less specialized ones.
+
+The list of specialization of the dists goes like follows:
+* sdist
+* py*-none-any.whl
+* py*-abi3-any.whl
+* py*-cpxy-any.whl
+* cp*-none-any.whl
+* cp*-abi3-any.whl
+* cp*-cpxy-plat.whl
+* py*-none-plat.whl
+* py*-abi3-plat.whl
+* py*-cpxy-plat.whl
+* cp*-none-plat.whl
+* cp*-abi3-plat.whl
+* cp*-cpxy-plat.whl
+
+Note, that here the specialization of musl vs manylinux wheels is the same in
+order to ensure that the matching fails if the user requests for `musl` and we don't have it or vice versa.
+"""
+
+load(":config_settings.bzl", "is_python_config_setting")
+load(
+    ":pip_flags.bzl",
+    "INTERNAL_FLAGS",
+    "UniversalWhlFlag",
+    "UseWhlFlag",
+    "WhlLibcFlag",
+)
+
+FLAGS = struct(
+    **{
+        f: str(Label("//python/config_settings:" + f))
+        for f in [
+            "python_version",
+            "pip_whl",
+            "pip_whl_glibc_version",
+            "pip_whl_muslc_version",
+            "pip_whl_osx_arch",
+            "pip_whl_osx_version",
+            "py_linux_libc",
+        ]
+    }
+)
+
+# Here we create extra string flags that are just to work with the select
+# selecting the most specialized match. We don't allow the user to change
+# them.
+_flags = struct(
+    **{
+        f: str(Label("//python/config_settings:_internal_pip_" + f))
+        for f in INTERNAL_FLAGS
+    }
+)
+
+def pip_config_settings(
```

Comment:
An idea from rules_pycross: https://github.com/jvolkman/rules_pycross/blob/531c8d21ff859e8b6df7f267eef066150723ec04/pycross/private/interpreter_version.bzl

Reply:
This could be a good alternative to the current `is_python_config_setting` implementation as it would reduce the number of flags that we generate.